modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+
+The syntax to define an object type with a set of constraints for its key value pairs, is very similar to defining an object value:
+
+----
+{
+    keyName @(attrName: AttrType): valueType,
+    ...
+}
+----
+
+For example, to define a user type with `firstName` String, `lastName` String and `age` Number, the syntax is:
+
+----
+type User = {
+    firstName: String,
+    lastName: String,
+    age: Number
+}
+----
+
+You can apply some modifiers to each key value pair to determine if a field is always present or if the field can be repeated or not.
+
+For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:
+
+----
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields use `?`. For example, the following type allows the age field to be present or not:
+
+----
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+==== Close and Open Objects
+
+Close object means that the type only accepts an object value if there are no additional fields in the object,
+just those key values specified by the type.
+
+Open object only puts constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accept objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
+
+The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:
+
+----
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a close object is `{| |}`. The following example, makes our user type only supports the keys `firstName` and `lastName`, so this throws an exception because of the `age` field is not accepted.
+
+----
+var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Function Type
+
+DataWeave is a functional language and functions are a first class citizen which means that functions are values
+and values have an associated type.
+
+The syntax to define a function type is:
+
+----
+(paramType: ParamType,...) -> ReturnType
+----
+
+For example, if you want to define a function that has a parameter of type String, a second parameter of type Number and return a Boolean, the correct syntax is:
+
+----
+(paramA: String, paramB: Number) -> Boolean
+----
+
+For example, you could define a function that receives another function as an argument:
+
+----
+fun findUserIndex(arr: Array<User>, condition: (userInArray: User, targetUser: User) -> Boolean): User = ???
+----
+
+If we call the function `findUserIndex` with some function that does not match the `condition` constraint `(User, User) -> Boolean`, DataWeave throws a type checking error.
```

Comment:
@TomoDC we could use examples here.

Reply:
`fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = 
    ids map changeTo($)`

`applyIDsChange([1,2,3], (item: Number) -> item as String)`

Maybe this example is simpler than the findUserIndex. The function call throws an exception because the function argument is of type (Number) -> String instead of (Number) -> (Number).

Reply:
@afelisatti @TomoDC let me know if you would like to change the function and text with the new sample Tomo provided in his last comment.

Reply:
Yep, I agree with adding the example.
`fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???`

`If we call the function **applyIDsChange** with some function that does not match the **changeTo** constraint **(Number) -> Number**, DataWeave throws a type checking error. For example in this function call **applyIDsChange([1,2,3], (item: Number) -> item as String)** DataWeave throws an exception because the function argument is of type (Number) -> String instead of (Number) -> Number.`

Reply:
Updated.